Indexing/SubstitutionTree.hpp

```diff
@@ -836,83 +1393,94 @@ class SubstitutionTree
   };  
 
   class UnificationsIterator
-  : public IteratorCore<QueryResult>
   {
   public:
-    UnificationsIterator(SubstitutionTree* parent, Node* root, Term* query, 
-      bool retrieveSubstitution, bool reversed, bool withoutTop, bool useC, 
-      FuncSubtermMap* funcSubtermMap = 0);
+    UnificationsIterator(UnificationsIterator&&) = default;
+    UnificationsIterator& operator=(UnificationsIterator&&) = default;
+    DECL_ELEMENT_TYPE(QueryResult);
+
+    template<class TermOrLit>
+    UnificationsIterator(SubstitutionTree* parent, Node* root, TermOrLit query, bool retrieveSubstitution, bool reversed, bool useC, FuncSubtermMap* funcSubtermMap)
+      : _subst()
+      , _svStack()
+      , _literalRetrieval(std::is_same<TermOrLit, Literal*>::value)
+      , _retrieveSubstitution(retrieveSubstitution)
+      , _inLeaf(false)
+      , _ldIterator(LDIterator::getEmpty())
+      , _nodeIterators()
+      , _bdStack()
+      , _clientBDRecording(false)
+      , _useUWAConstraints(useC)
+      , _useHOConstraints(funcSubtermMap)
+      , _constraints()
+      , _parentIterCntr(parent)
+#if VDEBUG
+      , _tag(parent->_tag)
+#endif
+    {
+#define DEBUG_QUERY(...) // DBG(__VA_ARGS__)
+      CALL("SubstitutionTree::UnificationsIterator::UnificationsIterator");
+
+      ASS(!_useUWAConstraints || retrieveSubstitution);
+      ASS(!_useUWAConstraints || parent->_useC);
+
+      if(!root) {
+        return;
+      }
+
+      if(funcSubtermMap){
+        _subst->setMap(funcSubtermMap);
+        query = ApplicativeHelper::replaceFunctionalAndBooleanSubterms(query, funcSubtermMap);
+      }
+
+      parent->createBindings(query, reversed, 
+          [&](unsigned var, TermList t) { _subst->bindSpecialVar(var, t, QUERY_BANK); });
+      DEBUG_QUERY("query: ", subst)
+
+
+      BacktrackData bd;
+      enter(root, bd);
+      bd.drop();
+    }
+
+
     ~UnificationsIterator();
 
     bool hasNext();
     QueryResult next();
-    bool tag;
   protected:
     virtual bool associate(TermList query, TermList node, BacktrackData& bd);
     virtual NodeIterator getNodeIterator(IntermediateNode* n);
 
-    void createInitialBindings(Term* t);
-    /**
-     * For a binary comutative literal, creates initial bindings,
-     * where the order of special variables is reversed.
-     */
-    void createReversedInitialBindings(Term* t);
     bool findNextLeaf();
     bool enter(Node* n, BacktrackData& bd);
 
 
     static const int QUERY_BANK=0;
     static const int RESULT_BANK=1;
-    static const int NORM_QUERY_BANK=2;
```

Comment:
Is `NORM_QUERY_BANK` unused? If so perhaps `NORM_RESULT_BANK` could equal 2?

Reply:
Yess true, because we don't normalize variable names of queries anymore (as this is unnecessary work that does not help us). 
I think I didn't change the constants because I'm a bit concerned that someone might check `if (bank == 3)` or similar somewhere and I didn'
t wanna look for bugs like this haha